File: go/lottery.go
package main

import (
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"time"
)

func play_game() int {
	var winning int
	var weights = []float64{1.0 / 302575351, 1.0 / 12607307, 1.0 / 931002,
		1.0 / 38793, 1.0 / 14548, 1.0 / 694, 1.0 / 607,
		1.0 / 90, 1.0 / 38, 24.0 / 25}
	var prize = []int{750000000, 1000000, 10000, 500, 200, 10, 10, 4, 2, 0}

	rnd := rand.Float64()
	for i, w := range weights {
		rnd -= w
		if rnd < 0 {
			winning = prize[i]
			break
		}
	}
	return winning
}

func saveResults(result []int) {
	var p string
	for _, elem := range result {
		p += strconv.Itoa(elem) + "\n"
	}

	err := os.WriteFile("go_results.txt", []byte(p), 0755)
	if err != nil {
		panic(err)
	}
}

func main() {
	rand.Seed(time.Now().UTC().UnixNano())
	var numOfSims, numOfGames = 50000, 5218
	fmt.Printf("Sims: %d - Games: %d\n", numOfSims, numOfGames)

	var gameResults = make([]int, numOfSims)
	var wins int

	for i, _ := range gameResults {
		wins = 0
		for j := 0; j < numOfGames; j++ {
			wins += play_game() - 2
		}
		gameResults[i] = wins
	}

	saveResults(gameResults)
}

File: c/lottery.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

static double runiform(void);
size_t weighted_random(const double weights[], const int length);
void mega_millions(int winnings[], const int numOfSims);
void save_results(const int winnings[], const int length );


int main() {
  const int NumberOfSims = 50000;
  int winnings[NumberOfSims];
  
  srand(time(NULL)); // random seed

  printf("Begin Mega Million simulation.\n");
  mega_millions(winnings, NumberOfSims);

  printf("Finished simulation, saving result.\n");
  save_results(winnings, NumberOfSims);

  return EXIT_SUCCESS;
}

// Returns uniform random var in [0,1].
// Used as a helper function in weighted_random.
double runiform(void) {
  double r = (double)rand() / RAND_MAX;
  return r;
}

// Returns an index based on the array weights.
// Assumes the weights sum to one.
size_t weighted_random(const double weights[], const int length) {
  double rnd = runiform();
  for (size_t i = 0; i < length; i++) {
    rnd -= weights[i];
    if (rnd < 0) {
      return i;
    }
  }
  return length+1; // outside of bounds when not matching
}

// playing the mega-millions
void mega_millions(int winnings[], const int numOfSims){
  double weights[10] = {1.0/302575351, 1.0/12607307, 1.0/931002,
                      1.0/38793, 1.0/14548, 1.0/694, 1.0/607,
                      1.0/90, 1.0/38, 24.0/25};

  int payout[10] = {750000000, 1000000, 10000, 500, 200, 10, 10, 4, 2, 0};
  
  int numOfGames = 5218;
  
  int wins;
  
  for (size_t i = 0; i < numOfSims; i++) {
    winnings[i] = 0; wins = 0;
    for (size_t j=0; j<numOfGames; j++){
      wins += payout[weighted_random(weights,10)] - 2;
    }
    winnings[i] = wins;
  } 
}

void save_results(const int winnings[], const int length ){
  FILE *fp = fopen("results.txt", "w");
  for (int i=0; i<length; i++){
    fprintf(fp, "%d\n", winnings[i]);
  }
  fclose(fp);
}